### PR TITLE
[#3757] Produce better information about partial parsing version mismatches

### LINKED ISSUE #3757: Unexpected occurrences of partial parsing version mismatch
From dbt community slack:

JC, I'm now using the 0.20.1 actual release and I'm still seeing occasions of Unable to do partial parsing because of a dbt version mismatch on runs.  It's not happening on every run, but it's definitely happening with regularity.
I'm having trouble generating a reproducible case (which is why I haven't opened an issue), but it's happening (with this project anyway) at least 1 in 3 runs.  I can't see anything that seems relevant in the logs, but I wanted to call it out in case (a) you hear of this elsewhere and (b) there's anything I can do to give you more context to understand why this might be happening.

Two consecutive commands were shared, both reporting the same version, with the second one claiming a version mismatch.

### PR DESCRIPTION
**Pull request by gshank**
resolves #3757

### Description

Gathers more information in order to debug spurious version mismatches which prevent partial parsing.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

File: core/dbt/parser/manifest.py [hunk 1]
```diff
@@ -493,7 +499,10 @@ def is_partial_parsable(self, manifest: Manifest) -> Tuple[bool, Optional[str]]:
         reparse_reason = None
 
         if manifest.metadata.dbt_version != __version__:
-            logger.info("Unable to do partial parsing because of a dbt version mismatch")
+            # #3757 log both versions because of reports of invalid cases of mismatch.
+            logger.info("Unable to do partial parsing because of a dbt version mismatch. "
+                        "Saved manifest version: {manifest.metadata.dbt_version}. "
+                        "Current version: {__version__}.")
```
Comment (leahwicz) [inline core/dbt/parser/manifest.py:505]: Does this work in Python without the `f` for inserting variables? I honestly don't know so asking for myself

**Review by leahwicz — APPROVED** [review]

File: core/dbt/parser/manifest.py [hunk 2]
```diff
@@ -478,6 +478,12 @@ def write_manifest_for_partial_parse(self):
         path = os.path.join(self.root_project.target_path,
                             PARTIAL_PARSE_FILE_NAME)
         try:
+            # This shouldn't be necessary, but we have gotten bug reports (#3757) of the
+            # saved manifest not matching the code version.
+            if self.manifest.metadata.dbt_version != __version__:
+                logger.debug("Manifest metadata did not contain correct version. "
+                             "Contained '{self.manifest.metadata.dbt_version}' instead.")
```
Comment (kwigley) [inline core/dbt/parser/manifest.py:485]:
```suggestion
                             f"Contained '{self.manifest.metadata.dbt_version}' instead.")
```

File: core/dbt/parser/manifest.py (hunk 1 above)
Comment (kwigley) [reply core/dbt/parser/manifest.py:505]: Ya, good call!

File: core/dbt/parser/manifest.py [hunk 3]
```diff
@@ -478,6 +478,12 @@ def write_manifest_for_partial_parse(self):
...
+                self.manifest.dbt_version = __version__
```
Comment (kwigley) [inline core/dbt/parser/manifest.py:486]: Should this be
```suggestion
                self.manifest.metadata.dbt_version = __version__
```

**Review by kwigley — DISMISSED** [review]

**Review by nathaniel-may — APPROVED** [review]

File: core/dbt/parser/manifest.py (hunk 1 above)
Comment (gshank) [reply core/dbt/parser/manifest.py:505]: Thanks!

File: core/dbt/parser/manifest.py (hunk 3 above)
Comment (gshank) [reply core/dbt/parser/manifest.py:486]: Arggh... Hard to test, but should have tried harder :)

**Comment by gshank** [comment]
Tested with 2 different version. Message: Unable to do partial parsing because of a dbt version mismatch. Saved manifest version: 0.20.1. Current version: 0.21.0b1.

**Pull request merged**